Hw8

chilik1020 opened the pull request.

chilik1020 committed:
Copied files from hw7

chilik1020 committed:
Added activities in manifest

chilik1020 committed:
Added ContactRepositoryCompletableFutureImpl

chilik1020 committed:
Added ContactRepositoryRxJavaImpl

chilik1020 committed:
Added BaseViewModel, EditContactViewModel

chilik1020 committed:
Added EditContact databinding, ContactAddViewModel, ContactsListViewModel

chilik1020 committed:
Added ContactAddBinding, ContactListBinding

chilik1020 committed:
Added filtering data in ContactsListViewModel

chilik1020 committed:
Added FetchInteractor, some package

chilik1020 committed:
Added CreateContactInteractor

chilik1020 committed:
Added EditContactUnteractor

chilik1020 committed:
Added WorkerHandlerThread, ContactRepositoryHandlerImpl

chilik1020 committed:
Added radio buttons binding in ContactAdd

chilik1020 committed:
added SettingsActivity

chilik1020 committed:
Fixed Settings icon position

DenisAndroidCource commented on hw8/src/main/java/com/chilik1020/hw8/model/repositories/ContactRepositoryCompletableFutureImpl.kt:
```diff
@@ -0,0 +1,50 @@
+        Log.d(LOG_TAG_APP, "CompletableFuture: getById")
+        val supplyAsyncById =
+            CompletableFuture.supplyAsync(Supplier<Contact> { contactDao.getById(id) }, executor)
+        listener.onSuccess(supplyAsyncById.get())
```
метод get является блокирующим. поэтому я бы посоветовал тут использовать runAsync и туда передать thread executor для главного потока, который можно попробовать передать либо в метод, либо в конструктор класса

DenisAndroidCource commented on hw8/src/main/java/com/chilik1020/hw8/model/repositories/ContactRepositoryCompletableFutureImpl.kt:
```diff
@@ -0,0 +1,50 @@
+        listener.onSuccess(supplyAsyncById.get())
+    }
+
+    override fun addContact(contact: Contact) {
+        Log.d(LOG_TAG_APP, "CompletableFuture: addContact")
+        CompletableFuture.runAsync(Runnable { contactDao.add(contact) }, executor)
+    }
+
+    override fun editContact(contact: Contact) {
+        Log.d(LOG_TAG_APP, "CompletableFuture: editContact")
+        CompletableFuture.runAsync(Runnable { contactDao.edit(contact) }, executor)
+    }
+
+    override fun removeContact(id: String) {
+        Log.d(LOG_TAG_APP, "CompletableFuture: removeContact")
+        CompletableFuture.runAsync(Runnable { contactDao.delete(id) }, executor)
+    }
```
кстати по-хорошему в таких операциях тоже лучше возвращать результат, потому что у тебя должна быть возможность обработать ошибку, если что-то пошло не так или обработать возможность реакции на операцию при успешном ее выполнении

chilik1020 committed:
Added .thenAcceptAsync(callback, mainThreadexecutor) in ContactRepositoryCompletableFutureImpl

chilik1020 committed:
Added return values in add, edit, delete methids in ContactDao

chilik1020 committed:
Added .thenAcceptAsync for add, edit,remove contact in CompletableFutureRepository

chilik1020 committed:
OnCreateContactListener in CreateContactInteractor

chilik1020 committed:
Added sealed class Result for quiery to db. Added SingleLiveEvent for Snackbar messages

chilik1020 committed:
Changed OnFetchContacrsListener to fun interface

chilik1020 committed:
Changed OnFetchContactByIdListener to fun interface

chilik1020 committed:
Added fun interfaces OnEditContactListener ,OnDeleteContactListener

chilik1020 committed:
.exceptionally() in CompletableFutureContactRepository

chilik1020 committed:
Added try-catch in HandlerContactRepository

DenisAndroidCource commented on hw8/src/main/java/com/chilik1020/hw8/model/repositories/ContactRepositoryCompletableFutureImpl.kt:
```diff
@@ -53,16 +53,38 @@ class ContactRepositoryCompletableFutureImpl(
+                mainThreadExecutor
+            )
     }
 }
```
Раз уж на то пошло, то стоит еще добавить функцию оставновки операций
